internal/watermill/middleware/dlq.go

```diff
@@ -0,0 +1,97 @@
+package middleware
+
+import (
+	"fmt"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+
+	"github.com/openmeterio/openmeter/internal/watermill"
+	"github.com/openmeterio/openmeter/pkg/errorsx"
+)
+
+const (
+	DLQErrorReasonKey   = "reason_dlq"
+	DLQOriginalTopicKey = "original_topic"
+)
+
+// DLQMsgProcessorFunc determines whether any messages should be published to the DLQ based on the original message and the error.
+// Even if DLQMsgProcessorFunc returns a non-nil slice of messages, if error is nil no messages will be published to the DLQ.
+type DLQMsgProcessorFunc func(original *message.Message, err error) []*message.Message
+
+// defaultProcessor listens for RetryableErrors and publishes the messages contained in that
+var defaultProcessor DLQMsgProcessorFunc = func(original *message.Message, err error) []*message.Message {
+	if retryable, ok := errorsx.ErrorAs[watermill.RetryableError](err); ok {
+		return retryable.RetryMessages()
+	}
+
+	return []*message.Message{original}
+}
+
+type deadLetterQueue struct {
+	topic string
+	pub   message.Publisher
+
+	msgProcessor DLQMsgProcessorFunc
+}
+
+// DLQ provides a middleware that salvages unprocessable messages and published them on a separate topic.
+// The main middleware chain then continues on, business as usual.
+func DLQ(pub message.Publisher, topic string, msgProcessor DLQMsgProcessorFunc) (message.HandlerMiddleware, error) {
```

Comment:
I don't think it's worth creating a DLQ implementation for this. I would rather have the following logic in the event handler:

```
- try to calculate the entitlements
- collect the successfully and unsuccessfully generated entitlements
- publish the successfully generated snapshots
- publish a EventBatchedIngest event with the remainder into the dlq topic
```

@sagikazarmark what do you think?

@GAlexIHU Given that I did a lot of refactoring around the handlers, I can take over this PR and wrap it up if you are fine with this.

Reply:
Sounds good to me, although I don't have a lot of context.

What you are suggesting sounds like "business logic" to me. DLQ is generally a good approach for blindly retrying without clogging up the original queue.

The above proposal sounds like a more "clever" retry approach where we only retry what needs to be retried. That sounds like business logic to me and DLQ may not be the best approach there.

Reply:
@turip I'm okay with you taking over, I've left this PR stale for the past some time no not much lost there

on the DLQ implementation topic, I think having a generic `RetryableError` or something of the sort will only make more sense as we go forward, but I'm fine either way for now